text/0000-normalized-modules.md

```diff
@@ -0,0 +1,535 @@
+        post-viewer
+          component.js
+          template.hbs
+      -helpers
+        capitalize.helper.js
+        titleize.helper.js
+      route.js
+      template.hbs
+  -components
+    list-paginator
+      component.js
+      template.hbs
+  -data
+    comments
+      adapter.js
+      model.js
+      serializer.js
+    posts
+      adapter.js
+      model.js
+      serializer.js
+  -helpers
+    capitalize.helper.js
+  -initializers
+    auth.instance-initializer.js
+    i18n.initializer.js
+  -services
+    auth.service.js
+  app.js
+  router.js
+  template.hbs
+```
+
+This example shows that we can improve the ergonomics of naming _files_ in
+Ember CLI projects, while normalizing everything into a single _module_
+naming structure.
+
+## Normalized module:file map
+
+The Ember CLI resolver will need to maintain a map of normalized module
+names to file locations on disk. This map will provide a single place for
+module lookups by the resolver.
+
+Construction of this map will allow for detection of any module name conflicts.
+In the case of any conflicts, Ember CLI should throw an exception.
+
+## Addon modules
+
+An addon's modules will no longer be "mixed in" to a consuming application
+directly. Instead, the addon's modules will be accessible at a namespace.
+Addons will be able to publish a default namespace (e.g. `ember-power-select`
+might choose `power-select`), which consuming applications will be able to
+alias to avoid conflicts.
+
+Private modules will exist alongside public modules in addons within a
+`-private` namespace. This namespace will provide a strong signal to consumers
+that private modules should not be accessed.
+
+### Namespaced components and helpers
+
+A `::` separator will be used to namespace components and helpers from
+templates.
+
+If an addon namespaced as `magic-ui` has a root-level `select/component`
+module, then that component can be rendered as `{{magic-ui::select}}`.
+
+Top-level component and helper modules named `main` can be accessed
+directly at the addon's namespace. If `ember-power-select` uses a top-level
+namespace of `power-select` and has a root-level `main/component` module, then
+that component can be rendered simply as `{{power-select}}`.
+
+## Local lookups
+
+Local lookup requires that module resolution occurs first "locally" before
+happening at the "global" (or top) level. This will improve the utility of
+components and helpers localized to an area of an application.
+
+Local lookups are made simple when there's only one place to look "locally" -
+in a namespace defined by the name of the current template.
+
+Say that the template `modules/posts/post-editor/template` references the
+`{{post-editor-button}}` component. This module could be found "locally" in
+`modules/posts/post-editor/post-editor-button/component`.
+
+## Co-existence with current module and file structures
+
+It will be necessary to allow a project to migrate its files from the current
```

Comment:
Is there plans to EOL existing project layouts in 3.0.0 or is the plan to support 3 project layouts?